enhancements/automated-service-ca-rotation.md

```diff
@@ -0,0 +1,246 @@
+
+## Motivation
+
+Service serving certificates are automatically renewed by the service CA
+controller. However, the service CA signer is not. The service CA is expired
+after 1 year, and once that expiration has passed clients will not be able to
+validate server certs, leading to service disruption. Recovering from this
+currently requires manual intervention (force rotation of the CA by deleting the
+CA secret).
+
+Since 4.1 shipped without automated rotation, operators depending on the service
+CA operator will break without manual triggering of forced rotation via deletion
+of the signing secret. Since manually-triggered rotation won't provide a grace
+period where the old and new CAs are trusted, it is likely to induce a brief
+period of system instability while servers and clients converge on the use of the
+new trust chain.
+
+### Goals
+
+- The signing CA should be automatically rotated before expiry
+
+- Rotation should prevent trust breakage between consumers of the operator as
+  long as the pre-rotation signing CA remains valid. The following should be
+  true:
+  - A serving cert, from the new CA, validates against the pre-rotation CA
+    bundle.
+  - A serving cert, from the old CA, validates against the post-rotation CA
+    bundle.
+
+- After rotation:
+  - Serving cert secrets should be updated with serving certs generated from the new CA.
+  - Configmaps annotated for CA bundle injection should be updated with the new CA bundle.
+  - API services annotated for CA bundle injection should be updated with the new CA bundle.
+
+- Support forced rotation
+  - Rotation triggered by an API change rather than the validity bounds of the CA
+
+- Support forced rotation with a custom validity bounds to support emergency rotation
+  - i.e. support minimizing the grace period for trusting the pre-rotation CA
+
+### Non-Goals
+
+- Support forced rotation using an external CA and signing key. This could be
+  useful for admins to plug in their own intermediate CA and have all serving
+  certs issued by it, and also for future integration efforts.
+
+- Implement support for [hitless
+rotation](https://diogomonica.com/2017/01/11/hitless-tls-certificate-rotation-in-go/)
+so that consumers of serving certs and CA bundles can trivially support automated
+refresh. This is already supported by
+[controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/pull/421)
+and the functionality could be separately extracted for operator reuse.
+
+## Proposal
+
+- Detect when the current CA certificate has passed 1/2 of its validity bounds
+  (i.e. 6 months has passed of a 1 year validity) and generate a new CA.
+  - The new CA should have the same `Subject.CommonName` as the current CA to
+    ensure that trust chaining is possible.
+  - This provides a 6 month grace period for consumers of serving certs and ca
+    bundles to be refreshed.
+
+- Generate an intermediate CA certificate with the same public key as the new CA
```

Comment:
the same public key and private key (this usually goes without saying since they are linked but I like to make sure it is explicit :))

Reply:
The private key is not relevant here because this intermediate CA cert will not be used to sign anything. It is only a bridge for trust.

Reply:
I didn't mean that the private key would be used, just nit-picking about the wording of this part :)

Reply:
I think the wording is correct. The private key is never signed since it is not part of the the 'tbs' (to-be-signed') portion of a certificate. A private key is only used for generating signatures of other certificates, and the trust of that private key is verified by decrypting a signature with a candidate public key and validating that the 'tbs' content of the signed cert matches.

Reply:
I think the wording is correct. The private key is never actually signed since it is not part of the the 'tbs' (to-be-signed) portion of a cert. 

I appreciate the question, though - answering it required me to learn more about how trust is verified. From my understanding, the tbs portion of a cert includes the public key, and the signature of a certificate is the hash of its tbs content encrypted by the private key of the issuer. To determine if a given certificate was issued by a candidate certificate:

- hash the tbs content of the certificate
- decrypt the signature of the certificate with the public key of the candidate certificate
- if the decrypted signature matches the certificate's tbs content, the certificate was issued by the candidate certificate